File: app/src/main/java/com/formichelli/dfnightselfies/preference/DFNightSelfiesPreferences.java
package com.formichelli.dfnightselfies.preference;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceActivity;
import android.view.ViewGroup;

import com.formichelli.dfnightselfies.R;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.kizitonwose.colorpreference.ColorPreference;

@SuppressWarnings("deprecation")
public class DFNightSelfiesPreferences extends PreferenceActivity implements SharedPreferences.OnSharedPreferenceChangeListener {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preferences);

        addAd();
    }

    protected void onResume() {
        super.onResume();

        for (String preferenceKey : getPreferenceScreen().getSharedPreferences().getAll().keySet()) {
            setSummary(preferenceKey);
        }

        getPreferenceScreen().getSharedPreferences()
                .registerOnSharedPreferenceChangeListener(this);
    }

    private void addAd() {
        // Create the adView
        AdView adView = new AdView(this);
        adView.setAdSize(AdSize.SMART_BANNER);
        adView.setAdUnitId(getString(R.string.banner_ad_unit_id));

        ((ViewGroup) findViewById(android.R.id.list).getParent().getParent().getParent()).addView(adView, 0);

        // Initiate a generic request to load it with an ad
        adView.loadAd(new AdRequest.Builder().addTestDevice("5B5C0102B231DC20553952DAC00561A6").build());
    }

    protected void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences()
                .unregisterOnSharedPreferenceChangeListener(this);
    }

    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences,
                                          String key) {
        setSummary(key);
    }

    private void setSummary(final String preferenceKey) {
        final Preference preference = findPreference(preferenceKey);

        if (preference instanceof CheckBoxPreference) {
            CheckBoxPreference cbp = (CheckBoxPreference) preference;
            preference.setSummary(getString(cbp.isChecked() ? R.string.enabled : R.string.disabled));
        } else if (preference instanceof ListPreference) {
            ListPreference lp = (ListPreference) preference;
            preference.setSummary(getString(R.string.countdown_preference_summary, Integer.valueOf(lp.getValue())));
        } else if (preference instanceof ColorPreference) {
            ColorPreference cp = (ColorPreference) preference;
            preference.setSummary(intToHexColor(cp));
        }
    }

    private String intToHexColor(ColorPreference cp) {
        return String.format("#%06X", (0xFFFFFF & cp.getValue()));
    }
}

File: app/src/main/java/com/formichelli/dfnightselfies/preference/DFNightSelfiesPreferences.kt
@file:Suppress("DEPRECATION")

package com.formichelli.dfnightselfies.preference

import android.content.SharedPreferences
import android.os.Bundle
import android.os.Environment
import android.preference.CheckBoxPreference
import android.preference.ListPreference
import android.preference.PreferenceActivity
import android.preference.SwitchPreference
import android.view.View
import android.view.ViewGroup
import com.formichelli.dfnightselfies.R
import com.formichelli.dfnightselfies.WindowUtil
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.kizitonwose.colorpreference.ColorPreference

class DFNightSelfiesPreferences : PreferenceActivity(), SharedPreferences.OnSharedPreferenceChangeListener {
    public override fun onCreate(savedInstanceState: Bundle?) {
        WindowUtil.setupWindow(this)

        super.onCreate(savedInstanceState)

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preferences)

        addAd()
    }

    override fun onResume() {
        super.onResume()

        preferenceScreen.sharedPreferences.all.keys.forEach {
            setSummary(it)
        }

        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    private fun addAd() {
        // Create the adView
        val adView = AdView(this)
        adView.adSize = AdSize.SMART_BANNER
        adView.adUnitId = getString(R.string.banner_ad_unit_id)

        (findViewById<View>(android.R.id.list).parent.parent.parent as ViewGroup).addView(adView, 0)

        // Initiate a generic request to load it with an ad
        adView.loadAd(AdRequest.Builder().addTestDevice("5B5C0102B231DC20553952DAC00561A6").build())
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) = setSummary(key)

    private fun setSummary(preferenceKey: String) {
        val preference = findPreference(preferenceKey)

        when (preference) {
            is CheckBoxPreference -> {
                if (preferenceKey == getString(R.string.save_to_gallery_preference)) {
                    val path = if (preferenceScreen.sharedPreferences.getBoolean(getString(R.string.save_to_gallery_preference), false)) {
                        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).absolutePath
                    } else {
                        Environment.getExternalStoragePublicDirectory(
                                Environment.DIRECTORY_PICTURES).absolutePath + "/" + getString(R.string.save_to_gallery_folder)
                    }

                    preference.setSummary(getString(R.string.save_to_gallery_summary, Environment.getExternalStoragePublicDirectory(
                            Environment.DIRECTORY_PICTURES).toString() + path))
                } else {
                    preference.setSummary(getString(if (preference.isChecked) R.string.enabled else R.string.disabled))
                }
            }
            is ListPreference -> {
                preference.setSummary(preference.value)
            }
            is ColorPreference -> {
                preference.setSummary(intToHexColor(preference))
            }
            is SwitchPreference -> {
                if (preferenceKey == getString(R.string.take_with_volume_preference)) {
                    val isEnabled = preferenceScreen.sharedPreferences.getBoolean(getString(R.string.take_with_volume_preference), false)
                    preference.setSummary(getString(if (isEnabled) R.string.take_with_volume_summary_true else R.string.take_with_volume_summary_false))
                } else {
                    preference.setSummary(getString(if (preference.isChecked) R.string.enabled else R.string.disabled))
                }
            }
        }
    }

    private fun intToHexColor(cp: ColorPreference) = String.format("#%06X", 0xFFFFFF and cp.value)
}
